### PR TITLE
feat(security): Implement spiffe token provider services

### PR DESCRIPTION
**Pull request by jim-wang-intel**
- add spire server Dockerfile
- add spire agent Dockerfile
- add spire config Dockerfile
- add spiffe token provider and `gettoken` restful endpoint

Closes: #3773

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

**If your build fails**  due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x ] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)    _spiffe/v2 and its related dependencies_ 
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)   x509Source from spiffe APIs not testable directly as they don't have interface to mock on
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?)  
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions

With updated docker-compose files from this branch (https://github.com/jim-wang-intel/edgex-compose/tree/add-spiffe-samples) and do `make run dev delayed-start` from the command line , then follow the testcase.txt under security-spiffe-token-provider/ service: https://github.com/jim-wang-intel/edgex-go/blob/add-spiffe-token-provider/cmd/security-spiffe-token-provider/testcase.txt 
to verify you can get a vault token.

## New Dependency Instructions (If applicable)

New dependency
- go-spiffe/v2
```
jim@yutsungj-DESK2:~/go/src/github.com/edgexfoundry/edgex-go$ GO111MODULE=on go list -m all | grep spiffe
github.com/spiffe/go-spiffe/v2 v2.0.0-beta.11
```

Total increase in new imports: 8
    Does the new import introduce additional import dependencies, if so, how many?
       Yes, totally **7** in addition to `go-spiffe/v2`
```
GO111MODULE=on go mod graph | grep spiffe
github.com/edgexfoundry/edgex-go github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/golang/protobuf@v1.4.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/stretchr/testify@v1.5.1
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 github.com/zeebo/errs@v1.2.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/grpc@v1.33.2
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/grpc/examples@v0.0.0-20201130180447-c456688b1860
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 google.golang.org/protobuf@v1.25.0
github.com/spiffe/go-spiffe/v2@v2.0.0-beta.11 gopkg.in/square/go-jose.v2@v2.4.1
```
    Releases/Tags: 19
    Contributors: 22
    License - what is the license, and is it Apache 2.0 compatible?  Apache License see https://github.com/spiffe/go-spiffe/blob/main/LICENSE 
    Stars/Forks/Watchers: 71 stars / 43 Forks / 20 Watchers
```
 godocs: https://pkg.go.dev/github.com/spiffe/go-spiffe/v2

### CODE REVIEW

**Comment by jim-wang-intel** [comment]
recheck

File: cmd/security-spiffe-token-provider/testcase.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+docker exec -ti edgex-security-test-client sh -l
+apk update && apk --no-cache --update add curl
+cd /tmp
+spire-agent api fetch x509 -socketPath /tmp/edgex/secrets/spiffe/public/api.sock -write /tmp
+curl -kiv -d service_key=security-test-client -d known_secret_names=redisdb --cert svid.0.pem --key svid.0.key "https://edgex-security-spiffe-token-provider:59841/api/v2/gettoken"
+curl -kiv -d service_key=security-test-client -d known_secret_names=redisdb --cert svid.0.pem --key svid.0.key "https://edgex-security-spiffe-token-provider:59841/api/v2/gettoken?raw_token=true"
```
Comment (bnevis-i) [inline cmd/security-spiffe-token-provider/testcase.txt:6]: Maybe move this into a README.md

File: cmd/security-spire-agent/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+COPY go.mod vendor* ./
+RUN [ ! -d "vendor" ] && go mod download all || echo "skipping..."
+
+COPY . .
+
+# The main mirrors are giving us timeout issues on builds periodically.
+# So we can try these.
+#RUN sed -e 's/dl-cdn[.]alpinelinux.org/nl.alpinelinux.org/g' -i~ /etc/apk/repositories
+#RUN apk add --update --no-cache zeromq-dev libsodium-dev pkgconfig build-base git
```
Comment (bnevis-i) [inline cmd/security-spire-agent/Dockerfile:36]: Remove these and add the `sed` command into the alpine container below prior to fetching gcompat.

File: cmd/security-spire-agent/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+ARG SPIRE_RELEASE=1.2.0
+
+RUN wget -q "https://github.com/spiffe/spire/releases/download/v${SPIRE_RELEASE}/spire-${SPIRE_RELEASE}-linux-x86_64-glibc.tar.gz" && \
+    tar xv --strip-components=1 -f "spire-${SPIRE_RELEASE}-linux-x86_64-glibc.tar.gz"
+
+# Deployment image
+#FROM debian:stable-slim
```
Comment (bnevis-i) [inline cmd/security-spire-agent/Dockerfile:44]: Remove.

File: cmd/security-spire-agent/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+FROM alpine:3.15
+
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2022 Intel Corporation'
+
+#ENV DEBIAN_FRONTEND=noninteractive
+#RUN apt-get update -y && apt-get upgrade -y && apt-get install -y --no-install-recommends dumb-init openssl && apt-get clean && rm -rf /var/lib/apt/lists/*
```
Comment (bnevis-i) [inline cmd/security-spire-agent/Dockerfile:51]: Remove.

File: cmd/security-spire-agent/docker-entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    echo "Waiting for /srv/spiffe/ca/public/agent-ca.crt"
+    sleep 1
+done
+
+# Create agent certificate
+
+if test ! -f "/srv/spiffe/agent/agent.crt"; then
+    openssl ecparam -genkey -name secp521r1 -noout -out "/srv/spiffe/agent/agent.key"
+    SAN="" openssl req -subj "/CN=${SPIFFE_AGENT0_CN}" -config "/usr/local/etc/openssl.conf" -key "/srv/spiffe/agent/agent.key" -sha384 -new -out "/run/agent.req.$$"
+    SAN="" openssl x509 -sha384 -extfile /usr/local/etc/openssl.conf -extensions agent_ext -CA "/srv/spiffe/ca/public/agent-ca.crt" -CAkey "/srv/spiffe/ca/private/agent-ca.key" -CAcreateserial -req -in "/run/agent.req.$$" -days 3650 -out "/srv/spiffe/agent/agent.crt"
```
Comment (bnevis-i) [inline cmd/security-spire-agent/docker-entrypoint.sh:53]: My bad.  Please bump these from sha384 to sha512.  (P-521 keys should use SHA-512 hashes).

File: cmd/security-spire-config/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+# Deployment image
+#FROM debian:stable-slim
+FROM alpine:3.15
+
+LABEL license='SPDX-License-Identifier: Apache-2.0' \
+      copyright='Copyright (c) 2022 Intel Corporation'
+
+#ENV DEBIAN_FRONTEND=noninteractive
+#RUN apt-get update -y && apt-get upgrade -y && apt-get install -y --no-install-recommends dumb-init && apt-get clean && rm -rf /var/lib/apt/lists/*
+RUN apk update && apk --no-cache --update add dumb-init gcompat
```
Comment (bnevis-i) [inline cmd/security-spire-config/Dockerfile:52]: See prior comments about removing the commented blocks now, and repeating the `sed` line prior to apk install for the alpine container.

File: cmd/security-spire-config/seed_builtin_entries.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+local_agent_svid=$1
+
+echo "local_agent_svid=${local_agent_svid}"
+echo "SPIFFE_SERVER_SOCKET=${SPIFFE_SERVER_SOCKET}"
+echo "SPIFFE_EDGEX_SVID_BASE=${SPIFFE_EDGEX_SVID_BASE}"
+
+# will replace security-test-client with all core-services later
+for service in security-spiffe-token-provider security-test-client; do
+    spire-server entry create -socketPath "${SPIFFE_SERVER_SOCKET}" -parentID "${local_agent_svid}" -dns "edgex-${service}" -spiffeID "${SPIFFE_EDGEX_SVID_BASE}/${service}" -selector "docker:label:com.docker.compose.service:${service}"
+done
```
Comment (bnevis-i) [inline cmd/security-spire-config/seed_builtin_entries.sh:31]: Whoops.  Add EOL

File: cmd/security-spire-server/docker-entrypoint.sh [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+           /srv/spiffe/ca/private ; do
+    test -d "$dir" || mkdir -p "$dir"
+done
+
+# CA SPIFFE identifiers
+
+if test ! -f "/srv/spiffe/ca/public/ca.crt"; then
+    openssl ecparam -genkey -name secp521r1 -noout -out "/srv/spiffe/ca/private/ca.key"
+    SAN="" openssl req -subj "/CN=SPIFFE Root CA" -config "/usr/local/etc/openssl.conf" -key "/srv/spiffe/ca/private/ca.key" -sha384 -new -out "/run/ca.req.$$"
+    SAN="" openssl x509 -sha384 -signkey "/srv/spiffe/ca/private/ca.key" -clrext -extfile /usr/local/etc/openssl.conf -extensions ca_ext -CAkey "/srv/spiffe/ca/private/ca.key" -CAcreateserial -req -in "/run/ca.req.$$" -days 3650 -out "/srv/spiffe/ca/public/ca.crt"
```
Comment (bnevis-i) [inline cmd/security-spire-server/docker-entrypoint.sh:44]: Bump to `-sha512`

File: go.mod [hunk 9]
```diff
@@ -14,6 +14,7 @@ require (
 	github.com/gorilla/mux v1.8.0
 	github.com/lib/pq v1.10.4
 	github.com/pelletier/go-toml v1.9.4
+	github.com/spiffe/go-spiffe/v2 v2.0.0-beta.11
```
Comment (bnevis-i) [inline go.mod:17]: AR: Collect the vetting data for the new dependency.

File: internal/security/spiffetokenprovider/init.go [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+	secretStoreClient, err := b.getSecretStoreClient(dic)
+	if err != nil {
+		lc.Errorf("failed to create SecretStoreClient: %s", err.Error())
+		return false
+	}
+
+	// Handle healthcheck endpoint
+	http.HandleFunc("/api/v2/ping", func(w http.ResponseWriter, r *http.Request) {
+		lc.Info("Request received for /api/v2/ping")
+		_, _ = io.WriteString(w, "pong")
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/init.go:138]: Add error handling.

File: internal/security/spiffetokenprovider/init.go [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+				// First certificate in verified chain has URI list,
+				// one of which is the SVID containing the EdgeX service key
+				if strings.HasPrefix(uri.String(), "spiffe://") {
+					peerSVID = uri.String()
+					break iterateCertificates
+				}
+			}
+		}
+
+		fmt.Println("PeerSVID:", peerSVID)
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/init.go:192]: Remove fmt.Println

File: internal/security/spiffetokenprovider/main.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+			"    --vaultInterval=<seconds>       Indicates how long the program will pause between vault initialization attempts until it succeeds",
+	)
+
+	if len(os.Args) < 2 {
+		f.Help()
+	}
+
+	f.FlagSet.BoolVar(&insecureSkipVerify, "insecureSkipVerify", false, "")
+	f.FlagSet.IntVar(&vaultInterval, "vaultInterval", 30, "")
+	f.Parse(os.Args[1:])
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/main.go:53]: I stole the wrong main.   These command line args are not appropriate.  Need a Main() that better resembles what a normal EdgeX service uses.

File: internal/security/spiffetokenprovider/main.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+		container.ConfigurationName: func(get di.Get) interface{} {
+			return configuration
+		},
+	})
+
+	bootstrap.Run(
+		ctx,
+		cancel,
+		f,
+		"security-spiffe-token-provider",
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/main.go:66]: Please start a PR for go-mod-core-contracts to add this as a new service key constant.

File: internal/security/spiffetokenprovider/init.go [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+		spiffeServiceKey := strings.Replace(peerSVID,
+			fmt.Sprintf("spiffe://%s/service/", configuration.Spiffe.TrustDomain.HostName), "", 1)
+
+		privilegedToken, err := b.getPrivilegedToken(dic)
+		if err != nil {
+			lc.Errorf("failed to load secret store token: %v", err)
+			w.WriteHeader(http.StatusForbidden)
+			return
+		}
+		vaultTokenResponse, err := makeToken(spiffeServiceKey, privilegedToken, secretStoreClient, lc)
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/init.go:204]: Before making the `makeToken` call, we need to apply the following logic from the ADR.  In the description below, "service name" is the service that was parsed from the SPIFFE ID using the regexp.  The "service key" is the form parameter that was passed by the client in the request.

---

The reqested service key. If blank, the service key will default to the service name encoded in the SVID. If the service name follows the pattern device-(name), then the service key must follow the format device-(name) or device-name-*. If the service name is app-service-configurable, then the service key must follow the format app-*. (This is an accomodation for the Unix workload attester not being able to distingish workloads that are launched using the same executable binary. Custom app services that support multiple instances won't be supported unless they name the executable the same as the standard app service binary or modify this logic.)

File: internal/security/spiffetokenprovider/init.go [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+	// attester not being able to distingish workloads that are launched using the same executable binary.
+	// Custom app services that support multiple instances won't be supported unless they name the executable
+	// the same as the standard app service binary or modify this logic.)
+	//
+	// A list of "known secret" identifiers that will allow new services to request database passwords or other
+	// "known secrets" to be seeded into their service's partition in the secret store.
+	//
+
+	// Handle gettoken endpoint
+	http.HandleFunc("/api/v2/gettoken", func(w http.ResponseWriter, r *http.Request) {
```
Comment (bnevis-i) [inline internal/security/spiffetokenprovider/init.go:159]: Need to limit this to HTTP POST.

File: Makefile [hunk 16]
```diff
@@ -8,6 +8,12 @@
 .PHONY: build clean unittest hadolint lint test docker run
 
 GO=CGO_ENABLED=0 GO111MODULE=on go
+
+# see https://shibumi.dev/posts/hardening-executables
+CGO_CPPFLAGS="-D_FORTIFY_SOURCE=2"
+CGO_CFLAGS="-O2 -pipe -fno-plt"
+CGO_CXXFLAGS="-O2 -pipe -fno-plt"
+CGO_LDFLAGS="-Wl,-O1,–sort-common,–as-needed,-z,relro,-z,now"
```
[the remark this answers is not in the document]
Comment (jim-wang-intel) [reply Makefile:16]: PR #3893

File: cmd/security-spire-agent/docker-entrypoint.sh [hunk 17]
```diff
@@ -50,7 +50,7 @@ done
 if test ! -f "/srv/spiffe/agent/agent.crt"; then
     openssl ecparam -genkey -name secp521r1 -noout -out "/srv/spiffe/agent/agent.key"
     SAN="" openssl req -subj "/CN=${SPIFFE_AGENT0_CN}" -config "/usr/local/etc/openssl.conf" -key "/srv/spiffe/agent/agent.key" -sha384 -new -out "/run/agent.req.$$"
-    SAN="" openssl x509 -sha384 -extfile /usr/local/etc/openssl.conf -extensions agent_ext -CA "/srv/spiffe/ca/public/agent-ca.crt" -CAkey "/srv/spiffe/ca/private/agent-ca.key" -CAcreateserial -req -in "/run/agent.req.$$" -days 3650 -out "/srv/spiffe/agent/agent.crt"
+    SAN="" openssl x509 -sha512 -extfile /usr/local/etc/openssl.conf -extensions agent_ext -CA "/srv/spiffe/ca/public/agent-ca.crt" -CAkey "/srv/spiffe/ca/private/agent-ca.key" -CAcreateserial -req -in "/run/agent.req.$$" -days 3650 -out "/srv/spiffe/agent/agent.crt"
```
Comment (bnevis-i) [inline cmd/security-spire-agent/docker-entrypoint.sh:53]: Missed -sha512 on the line above.

File: cmd/security-spire-server/docker-entrypoint.sh [hunk 18]
```diff
@@ -41,7 +41,7 @@ done
 if test ! -f "/srv/spiffe/ca/public/ca.crt"; then
     openssl ecparam -genkey -name secp521r1 -noout -out "/srv/spiffe/ca/private/ca.key"
     SAN="" openssl req -subj "/CN=SPIFFE Root CA" -config "/usr/local/etc/openssl.conf" -key "/srv/spiffe/ca/private/ca.key" -sha384 -new -out "/run/ca.req.$$"
-    SAN="" openssl x509 -sha384 -signkey "/srv/spiffe/ca/private/ca.key" -clrext -extfile /usr/local/etc/openssl.conf -extensions ca_ext -CAkey "/srv/spiffe/ca/private/ca.key" -CAcreateserial -req -in "/run/ca.req.$$" -days 3650 -out "/srv/spiffe/ca/public/ca.crt"
+    SAN="" openssl x509 -sha512 -signkey "/srv/spiffe/ca/private/ca.key" -clrext -extfile /usr/local/etc/openssl.conf -extensions ca_ext -CAkey "/srv/spiffe/ca/private/ca.key" -CAcreateserial -req -in "/run/ca.req.$$" -days 3650 -out "/srv/spiffe/ca/public/ca.crt"
```
Comment (bnevis-i) [inline cmd/security-spire-server/docker-entrypoint.sh:44]: Missed -sha512 on the line above.

File: cmd/security-spire-server/docker-entrypoint.sh [hunk 19]
```diff
@@ -50,7 +50,7 @@ fi
 if test ! -f "/srv/spiffe/ca/public/agent-ca.crt"; then
     openssl ecparam -genkey -name secp521r1 -noout -out "/srv/spiffe/ca/private/agent-ca.key"
     SAN="" openssl req -subj "/CN=SPIFFE Agent CA" -config "/usr/local/etc/openssl.conf" -key "/srv/spiffe/ca/private/agent-ca.key" -sha384 -new -out "/run/ca.req.$$"
-    SAN="" openssl x509 -sha384 -signkey "/srv/spiffe/ca/private/agent-ca.key" -clrext -extfile /usr/local/etc/openssl.conf -extensions ca_ext -CAkey "/srv/spiffe/ca/private/agent-ca.key" -CAcreateserial -req -in "/run/ca.req.$$" -days 3650 -out "/srv/spiffe/ca/public/agent-ca.crt"
+    SAN="" openssl x509 -sha512 -signkey "/srv/spiffe/ca/private/agent-ca.key" -clrext -extfile /usr/local/etc/openssl.conf -extensions ca_ext -CAkey "/srv/spiffe/ca/private/agent-ca.key" -CAcreateserial -req -in "/run/ca.req.$$" -days 3650 -out "/srv/spiffe/ca/public/agent-ca.crt"
```
Comment (bnevis-i) [inline cmd/security-spire-server/docker-entrypoint.sh:53]: Missed -sha512 on the line above.

File: internal/security/spiffetokenprovider/main.go (hunk 13 above)
Comment (jim-wang-intel) [reply internal/security/spiffetokenprovider/main.go:83]: PR 702  in  https://github.com/edgexfoundry/go-mod-core-contracts/pull/702

File: cmd/security-spiffe-token-provider/res/configuration.toml [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+RootCaCertPath = ""
+ServerName = ""
+  [SecretStore.Authentication]
+  AuthType = "X-Vault-Token"
+
+[SPIFFE]
+  [SPIFFE.Endpoint]
+  Socket = "/tmp/edgex/secrets/spiffe/public/api.sock"
+  [SPIFFE.TrustDomain]
+  HostName = "edgexfoundry.org"
```
Comment (bnevis-i) [inline cmd/security-spiffe-token-provider/res/configuration.toml:58]: I think we need to eliminate the redundancy here.  We have

SPIFFE_ENDPOINT_SOCKET <https://github.com/spiffe/go-spiffe/tree/main/v2#quick-start>

But I think it's a mistake to have

SPIFFE_TRUST_DOMAIN and SPIFFE_TRUSTDOMAIN_HOSTNAME

Let's do one or the other.

**Review by bnevis-i — APPROVED** [review]

**Comment by jim-wang-intel** [comment]
Close this PR, duplicated changed from the latest new PR #3897

**Pull request closed without merging**